kv/kvclient/kvcoord: TestTxnCoordSenderRetries failed

[(kv/kvclient/kvcoord).TestTxnCoordSenderRetries failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1877240&tab=buildLog) on [master@8c928666de3f90cf17d81d70a3ee759c39e2ce27](https://github.com/cockroachdb/cockroach/commits/8c928666de3f90cf17d81d70a3ee759c39e2ce27):

```
=== RUN   TestTxnCoordSenderRetries
I200415 05:43:47.761583 39254 base/addr_validation.go:296  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I200415 05:43:47.761609 39254 base/addr_validation.go:342  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
W200415 05:43:47.761732 39254 server/status/runtime.go:308  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I200415 05:43:47.771834 39254 server/server.go:1229  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I200415 05:43:47.775684 39254 server/config.go:576  [n?] 1 storage engine initialized
I200415 05:43:47.775716 39254 server/config.go:579  [n?] RocksDB cache size: 128 MiB
I200415 05:43:47.775726 39254 server/config.go:579  [n?] store 0: in-memory, size 0 B
I200415 05:43:47.780455 39254 server/server.go:1749  [n?] **** add additional nodes by specifying --join=127.0.0.1:39651
I200415 05:43:47.780491 39254 server/init.go:166  [n?] no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.
I200415 05:43:47.780508 39254 server/init.go:186  [n?] **** cluster 8287d6a9-dc45-4cdc-8ac2-704d52316531 has been created
I200415 05:43:47.780602 39254 gossip/gossip.go:404  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39651" > attrs:<> locality:<> ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:1 > build_tag:"v20.1.0-beta.4-759-g8c92866" started_at:1586929427780545026 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:45465" > 
I200415 05:43:47.780661 39428 server/server.go:1816  [n1] connecting to gossip network to verify cluster ID "8287d6a9-dc45-4cdc-8ac2-704d52316531"
W200415 05:43:47.783871 39503 kv/kvserver/store.go:1627  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200415 05:43:47.783882 39428 server/server.go:1819  [n1] node connected via gossip
I200415 05:43:47.784209 39254 server/node.go:416  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=16 KiB), ranges=32, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=166.00 pMax=13304.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I200415 05:43:47.784311 39254 kv/kvserver/stores.go:236  [n1] read 0 node addresses from persistent storage
I200415 05:43:47.784376 39254 server/node.go:472  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200415 05:43:47.784456 39254 server/server.go:1913  [n1] starting https server at 127.0.0.1:45719 (use: 127.0.0.1:45719)
I200415 05:43:47.784491 39254 server/server.go:1918  [n1] starting postgres server at 127.0.0.1:45465 (use: 127.0.0.1:45465)
I200415 05:43:47.784508 39254 server/server.go:1920  [n1] starting grpc server at 127.0.0.1:39651
I200415 05:43:47.784529 39254 server/server.go:1921  [n1] advertising CockroachDB node at 127.0.0.1:39651
I200415 05:43:47.784690 39524 sql/temporary_schema.go:459  [n1] running temporary object cleanup background job
I200415 05:43:47.795654 39384 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200415 05:43:47.800919 39715 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1-1 User:root}
I200415 05:43:47.811460 39975 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:fa8abce4-dc3c-41ca-ad09-e2ed1c820f12 User:root}
I200415 05:43:47.815993 39733 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200415 05:43:47.819214 39870 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200415 05:43:47.831632 39254 server/server.go:2075  [n1] done ensuring all necessary migrations have run
I200415 05:43:47.831663 39254 server/server.go:2311  [n1] serving sql connections
I200415 05:43:47.831730 39983 sqlmigrations/migrations.go:663  [n1] starting wait for upgrade finalization before schema change job migration
I200415 05:43:47.831959 39983 sqlmigrations/migrations.go:695  [n1] starting schema change job migration
I200415 05:43:47.832946 39254 kv/kvserver/replica_command.go:397  [n1,s1,r5/1:/{Systemtse-Table/System…}] initiating a split of this range at key "b" [r33] (manual)
I200415 05:43:47.834654 39981 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200415 05:43:47.835495 39983 sqlmigrations/migrations.go:933  [schema-change-job-migration] evaluating tables for creating jobs
I200415 05:43:47.835528 39983 sqlmigrations/migrations.go:714  [n1] schema change job migration completed
I200415 05:43:47.836494 39984 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:39651 Attrs: Locality: ServerVersion:20.1-1 BuildTag:v20.1.0-beta.4-759-g8c92866 StartedAt:1586929427780545026 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:45465} ClusterID:8287d6a9-dc45-4cdc-8ac2-704d52316531 StartedAt:1586929427780545026 LastUp:1586929427780545026}
--- FAIL: TestTxnCoordSenderRetries (0.20s)
=== RUN   TestTxnCoordSenderRetries/forwarded_timestamp_with_get_and_cput_after_timestamp_leaked
I200415 05:43:47.841932 39524 sql/temporary_schema.go:494  [n1] found 0 temporary schemas
I200415 05:43:47.841961 39524 sql/temporary_schema.go:497  [n1] early exiting temporary schema cleaner as no temporary schemas were found
I200415 05:43:47.841977 39524 sql/temporary_schema.go:498  [n1] completed temporary object cleanup job
I200415 05:43:47.842000 39524 sql/temporary_schema.go:575  [n1] temporary object cleaner next scheduled to run at 2020-04-15 06:13:47.784633517 +0000 UTC
    --- FAIL: TestTxnCoordSenderRetries/forwarded_timestamp_with_get_and_cput_after_timestamp_leaked (0.00s)
        dist_sender_server_test.go:2875: expected no txn coord sender auto retries; got 1

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestTxnCoordSenderRetries PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTxnCoordSenderRetries.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvclient/kvcoord).TestTxnCoordSenderRetries failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1879767&tab=buildLog) on [master@de48c6d13431061a6227697a917e574464b80983](https://github.com/cockroachdb/cockroach/commits/de48c6d13431061a6227697a917e574464b80983):

```
=== RUN   TestTxnCoordSenderRetries
I200416 06:17:59.402967 38794 base/addr_validation.go:296  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I200416 06:17:59.403004 38794 base/addr_validation.go:342  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
W200416 06:17:59.403097 38794 server/status/runtime.go:308  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I200416 06:17:59.425804 38794 server/server.go:1229  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I200416 06:17:59.436554 38794 server/config.go:576  [n?] 1 storage engine initialized
I200416 06:17:59.436574 38794 server/config.go:579  [n?] RocksDB cache size: 128 MiB
I200416 06:17:59.436582 38794 server/config.go:579  [n?] store 0: in-memory, size 0 B
I200416 06:17:59.440550 38794 server/server.go:1749  [n?] **** add additional nodes by specifying --join=127.0.0.1:40949
I200416 06:17:59.440573 38794 server/init.go:166  [n?] no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.
I200416 06:17:59.440583 38794 server/init.go:186  [n?] **** cluster 7353dfa9-0e5e-40a4-b652-4d2f096584b3 has been created
I200416 06:17:59.440670 38794 gossip/gossip.go:404  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:40949" > attrs:<> locality:<> ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:1 > build_tag:"v20.1.0-beta.4-778-gde48c6d" started_at:1587017879440626448 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:35093" > 
I200416 06:17:59.444310 38855 server/server.go:1816  [n1] connecting to gossip network to verify cluster ID "7353dfa9-0e5e-40a4-b652-4d2f096584b3"
I200416 06:17:59.444347 38794 server/node.go:416  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=16 KiB), ranges=32, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=166.00 pMax=13304.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I200416 06:17:59.444415 38794 kv/kvserver/stores.go:236  [n1] read 0 node addresses from persistent storage
W200416 06:17:59.444652 38930 kv/kvserver/store.go:1627  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200416 06:17:59.444685 38794 server/node.go:472  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200416 06:17:59.444541 38855 server/server.go:1819  [n1] node connected via gossip
I200416 06:17:59.444770 38794 server/server.go:1913  [n1] starting https server at 127.0.0.1:41975 (use: 127.0.0.1:41975)
I200416 06:17:59.444803 38794 server/server.go:1918  [n1] starting postgres server at 127.0.0.1:35093 (use: 127.0.0.1:35093)
I200416 06:17:59.444817 38794 server/server.go:1920  [n1] starting grpc server at 127.0.0.1:40949
I200416 06:17:59.444831 38794 server/server.go:1921  [n1] advertising CockroachDB node at 127.0.0.1:40949
I200416 06:17:59.444913 38783 sql/temporary_schema.go:459  [n1] running temporary object cleanup background job
I200416 06:17:59.456237 38968 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200416 06:17:59.461966 38811 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1-1 User:root}
I200416 06:17:59.471001 39422 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:290f203c-198b-4401-a0d0-9072b8476c33 User:root}
I200416 06:17:59.474657 39279 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200416 06:17:59.477746 39523 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200416 06:17:59.489555 38794 server/server.go:2075  [n1] done ensuring all necessary migrations have run
I200416 06:17:59.489582 38794 server/server.go:2311  [n1] serving sql connections
I200416 06:17:59.489634 39687 sqlmigrations/migrations.go:663  [n1] starting wait for upgrade finalization before schema change job migration
I200416 06:17:59.489831 39687 sqlmigrations/migrations.go:695  [n1] starting schema change job migration
I200416 06:17:59.490716 38794 kv/kvserver/replica_command.go:397  [n1,s1,r5/1:/{Systemtse-Table/System…}] initiating a split of this range at key "b" [r33] (manual)
I200416 06:17:59.491288 39688 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:40949 Attrs: Locality: ServerVersion:20.1-1 BuildTag:v20.1.0-beta.4-778-gde48c6d StartedAt:1587017879440626448 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:35093} ClusterID:7353dfa9-0e5e-40a4-b652-4d2f096584b3 StartedAt:1587017879440626448 LastUp:1587017879440626448}
I200416 06:17:59.491645 39685 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200416 06:17:59.493272 39687 sqlmigrations/migrations.go:933  [schema-change-job-migration] evaluating tables for creating jobs
I200416 06:17:59.493301 39687 sqlmigrations/migrations.go:714  [n1] schema change job migration completed
--- FAIL: TestTxnCoordSenderRetries (0.21s)
=== RUN   TestTxnCoordSenderRetries/forwarded_timestamp_with_scan_and_cput
I200416 06:17:59.500485 38783 sql/temporary_schema.go:494  [n1] found 0 temporary schemas
I200416 06:17:59.500509 38783 sql/temporary_schema.go:497  [n1] early exiting temporary schema cleaner as no temporary schemas were found
I200416 06:17:59.500523 38783 sql/temporary_schema.go:498  [n1] completed temporary object cleanup job
I200416 06:17:59.500540 38783 sql/temporary_schema.go:575  [n1] temporary object cleaner next scheduled to run at 2020-04-16 06:47:59.44481428 +0000 UTC
    --- FAIL: TestTxnCoordSenderRetries/forwarded_timestamp_with_scan_and_cput (0.00s)
        dist_sender_server_test.go:2875: expected no txn coord sender auto retries; got 1

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestTxnCoordSenderRetries PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTxnCoordSenderRetries.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
=== RUN   TestTxnCoordSenderRetries/write_too_old_with_locking_read
    --- FAIL: TestTxnCoordSenderRetries/write_too_old_with_locking_read (0.01s)
        dist_sender_server_test.go:2860: cannot read undeclared span a�
            declared:
            read local: /Local/RangeID/5/r/AbortSpan/"393db0ee-7ace-43e7-b459-1171d0fd2734" at 0,0
            write global: a{-�} at 1589060540.751137798,0
            
            stack:
            goroutine 40710 [running]:
            runtime/debug.Stack(0xc00426e6d8, 0x1, 0x8)
            	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
            github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).checkAllowed(0xc0022ff6e0, 0x0, 0xc00426e930, 0x2, 0x8, 0x0, 0x0, 0x0, 0xc003abaeb0, 0xc001fef080, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:293 +0x3ad
            github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).CheckAllowedAt(0xc0022ff6e0, 0x0, 0xc00426e930, 0x2, 0x8, 0x0, 0x0, 0x0, 0x160d7a24820fcc06, 0x0, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:235 +0x116
            github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*Iterator).SeekGE(0xc004389b80, 0xc00426e930, 0x2, 0x8, 0x0, 0x0)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:74 +0x367
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).iterSeek(0xc001e6a000, 0xc00426e930, 0x2, 0x8, 0x0, 0x0, 0x1)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:636 +0xb4
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).nextKey(0xc001e6a000, 0xc00426e750)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:456 +0x57e
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).advanceKey(0xc001e6a000, 0xc00426e750)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:517 +0xdc
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).addAndAdvance(0xc001e6a000, 0x7fd9ffcf920b, 0x8, 0x8, 0x1)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:569 +0x6b
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).getAndAdvance(0xc001e6a000, 0xc00426e6d8)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:274 +0x53e
            github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).scan(0xc001e6a000, 0x10, 0xc001e6a000, 0x7fda0d89cc08)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:180 +0x167
            github.com/cockroachdb/cockroach/pkg/storage.mvccScanToBytes(0x6daf3a0, 0xc002d88600, 0x6e36b40, 0xc004389b80, 0xc00426e6d8, 0x1, 0x8, 0xc00426e6f0, 0x2, 0x8, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2366 +0x48a
            github.com/cockroachdb/cockroach/pkg/storage.mvccScanToKvs(0x6daf3a0, 0xc002d88600, 0x6e36b40, 0xc004389b80, 0xc00426e6d8, 0x1, 0x8, 0xc00426e6f0, 0x2, 0x8, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2396 +0x13d
            github.com/cockroachdb/cockroach/pkg/storage.MVCCScan(0x6daf3a0, 0xc002d88600, 0x6df69e0, 0xc0022ff7a0, 0xc00426e6d8, 0x1, 0x8, 0xc00426e6f0, 0x2, 0x8, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2539 +0x276
            github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.Scan(0x6daf3a0, 0xc002d88600, 0x6df69e0, 0xc0022ff7a0, 0x6e47b40, 0xc004146230, 0x160d7a2482169573, 0x1, 0x100000001, 0x1, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_scan.go:60 +0x984
            github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x6daf3a0, 0xc002d88600, 0x0, 0x0, 0x0, 0x6e354c0, 0xc0022ff7a0, 0x6e47b40, 0xc004146230, 0x0, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:473 +0x74c
            github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x6daf3a0, 0xc002d88600, 0x0, 0x0, 0x6e354c0, 0xc0022ff7a0, 0x6e47b40, 0xc004146230, 0x0, 0xc003502100, ...)
            	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:241 +0x492
            github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatchWithServersideRefreshes(0xc000cfa800, 0x6daf3a0, 0xc002d88600, 0x6e354c0, 0xc0022ff7a0, 0x6e47b40, 0xc004146230, 0xc003502100, 0xc0022ff6e0, 0xc005df6178, ...)
```
This is due to the switch to Pebble and is similar to the failure fixed by https://github.com/cockroachdb/cockroach/pull/48490. The fix will probably be similar as well.

--- a/pkg/kv/kvserver/batch_spanset_test.go
+++ b/pkg/kv/kvserver/batch_spanset_test.go
@@ -223,7 +223,7 @@ func TestSpanSetBatchBoundaries(t *testing.T) {
 	t.Run("reverse scans", func(t *testing.T) {
 		iter := spanset.NewIterator(eng.NewIterator(storage.IterOptions{UpperBound: roachpb.KeyMax}), &ss)
 		defer iter.Close()
-		iter.SeekLT(outsideKey)
+		iter.SeekLT(outsideKey4)
 		if _, err := iter.Valid(); !isReadSpanErr(err) {
 			t.Fatalf("SeekLT: unexpected error %v", err)
 		}

--- a/pkg/kv/kvserver/spanset/batch.go
+++ b/pkg/kv/kvserver/spanset/batch.go
@@ -66,35 +66,6 @@ func (i *Iterator) Iterator() storage.Iterator {
 	return i.i
 }
 
-// SeekGE is part of the engine.Iterator interface.
-func (i *Iterator) SeekGE(key storage.MVCCKey) {
-	if i.spansOnly {
-		i.err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: key.Key})
-	} else {
-		i.err = i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: key.Key}, i.ts)
-	}
-	if i.err == nil {
-		i.invalid = false
-	}
-	i.i.SeekGE(key)
-}
-
-// SeekLT is part of the engine.Iterator interface.
-func (i *Iterator) SeekLT(key storage.MVCCKey) {
-	// CheckAllowed{At} supports the span representation of [,key), which
-	// corresponds to the span [key.Prev(),).
-	revSpan := roachpb.Span{EndKey: key.Key}
-	if i.spansOnly {
-		i.err = i.spans.CheckAllowed(SpanReadOnly, revSpan)
-	} else {
-		i.err = i.spans.CheckAllowedAt(SpanReadOnly, revSpan, i.ts)
-	}
-	if i.err == nil {
-		i.invalid = false
-	}
-	i.i.SeekLT(key)
-}
-
 // Valid is part of the engine.Iterator interface.
 func (i *Iterator) Valid() (bool, error) {
 	if i.err != nil {
@@ -107,51 +78,57 @@ func (i *Iterator) Valid() (bool, error) {
 	return ok && !i.invalid, nil
 }
 
+// SeekGE is part of the engine.Iterator interface.
+func (i *Iterator) SeekGE(key storage.MVCCKey) {
+	i.i.SeekGE(key)
+	i.checkAllowed(roachpb.Span{Key: key.Key}, true)
+}
+
+// SeekLT is part of the engine.Iterator interface.
+func (i *Iterator) SeekLT(key storage.MVCCKey) {
+	i.i.SeekLT(key)
+	// CheckAllowed{At} supports the span representation of [,key), which
+	// corresponds to the span [key.Prev(),).
+	i.checkAllowed(roachpb.Span{EndKey: key.Key}, true)
+}
+
 // Next is part of the engine.Iterator interface.
 func (i *Iterator) Next() {
 	i.i.Next()
-	if ok, _ := i.i.Valid(); ok {
-		if i.spansOnly {
-			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-				i.invalid = true
-			}
-		} else {
-			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-				i.invalid = true
-			}
-		}
-	}
+	i.checkAllowed(roachpb.Span{Key: i.UnsafeKey().Key}, false)
 }
 
 // Prev is part of the engine.Iterator interface.
 func (i *Iterator) Prev() {
 	i.i.Prev()
-	if ok, _ := i.i.Valid(); ok {
-		if i.spansOnly {
-			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-				i.invalid = true
-			}
-		} else {
-			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-				i.invalid = true
-			}
-		}
-	}
+	i.checkAllowed(roachpb.Span{Key: i.UnsafeKey().Key}, false)
 }
 
 // NextKey is part of the engine.Iterator interface.
 func (i *Iterator) NextKey() {
 	i.i.NextKey()
-	if ok, _ := i.i.Valid(); ok {
-		if i.spansOnly {
-			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-				i.invalid = true
-			}
-		} else {
-			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-				i.invalid = true
-			}
-		}
+	i.checkAllowed(roachpb.Span{Key: i.UnsafeKey().Key}, false)
+}
+
+func (i *Iterator) checkAllowed(span roachpb.Span, errIfDisallowed bool) {
+	i.invalid = false
+	i.err = nil
+	if ok, _ := i.i.Valid(); !ok {
+		// If the iterator is invalid after the operation, there's nothing to
+		// check. We allow uses of iterators to exceed the declared span bounds
+		// as long as the iterator itself is configured with proper boundaries.
+		return
+	}
+	var err error
+	if i.spansOnly {
+		err = i.spans.CheckAllowed(SpanReadOnly, span)
+	} else {
+		err = i.spans.CheckAllowedAt(SpanReadOnly, span, i.ts)
+	}
+	if errIfDisallowed {
+		i.err = err
+	} else {
+		i.invalid = err != nil
 	}
 }
 

